sherpa/astro/instrument.py

```diff
@@ -1141,3 +1142,120 @@ def get_kernel(self, data, subkernel=True):
+                     name='delta-rmf'):
+
+    _validate_create_rmf_input(rmflo, rmfhi, offset)
+
+    # Set up the delta-function response.
+    # TODO: should f_chan start at startchan?
+    #
+    nchans = rmflo.size
+    matrix = numpy.ones(nchans, dtype=numpy.float32)
+    dummy = numpy.ones(nchans, dtype=numpy.int16)
+    f_chan = numpy.arange(1, nchans + 1, dtype=numpy.int16)
+
+    return sherpa.astro.data.DataRMF(name, detchans=nchans,
+                                     energ_lo=rmflo, energ_hi=rmfhi,
+                                     n_grp=dummy, n_chan=dummy,
+                                     f_chan=f_chan, matrix=matrix,
+                                     offset=offset,
+                                     e_min=e_min, e_max=e_max,
+                                     ethresh=ethresh)
+
+
+def create_non_delta_rmf(rmflo, rmfhi, fname, offset=1,
+                         e_min=None, e_max=None, ethresh=None,
+                         name='delta-rmf'):
+
+    if fname is not None and not os.path.isfile(fname):
+        raise ValueError("{} is not a file".format(fname))
+
+    # Set up the delta-function response.
+    # TODO: should f_chan start at startchan?
+    #
+    nchans = rmflo.size
+
+    n_grp, f_chan, n_chan, matrix = calc_grp_chan_matrix(fname)
+
+    return sherpa.astro.data.DataRMF(name, detchans=nchans,
+                                     energ_lo=rmflo, energ_hi=rmfhi,
+                                     n_grp=n_grp, n_chan=n_chan,
+                                     f_chan=f_chan, matrix=matrix,
+                                     offset=offset,
+                                     e_min=e_min, e_max=e_max,
+                                     ethresh=ethresh)
+
+
+def calc_grp_chan_matrix(fname):
+    try:
+        data, filename = \
+            sherpa.astro.io.backend.get_image_data(fname)
+        matrix = data['y']
+        n_grp = []
+        n_chan = []
+        f_chan = []
+        for row in matrix > 0:
+            flag = numpy.hstack([[0], row, [0]])
+            diffs = numpy.diff(flag, n=1)
+            starts, = numpy.where(diffs > 0)
+            ends, = numpy.where(diffs < 0)
+            n_chan.extend(ends - starts)
+            f_chan.extend(starts + 1)
+            n_grp.append(len(starts))
+        n_grp = numpy.asarray(n_grp, dtype=numpy.int16)
+        f_chan = numpy.asarray(f_chan, dtype=numpy.int16)
+        n_chan = numpy.asarray(n_chan, dtype=numpy.int16)
+        matrix = matrix.flatten()
+        matrix = matrix[matrix > 0]
+        return n_grp, f_chan, n_chan, matrix
+    except sherpa.utils.err.IOErr as ioerr:
+        print(ioerr)
+        raise ioerr
+
+
+def _validate_create_rmf_input(rmflo, rmfhi, startchan, fname=None):
+    try:
+        assert rmflo.size == rmfhi.size
+    except AssertionError:
+        raise ValueError("rmflo and rmfhi must have the same size, not {} and {}".format(rmflo.size, rmfhi.size))
+
+    try:
```

Comment:
Why not just 'if startchan < 0: raise ValueError(...)' ?

Reply:
And similarly the check on `rmflo.size == rmfhi.size`?

Reply:
Yes, that's part of what I wanted to avoid with the refactoring. Thanks.